backends/simplego/dotgeneral.go

```diff
@@ -63,9 +63,21 @@ func (b *Builder) DotGeneral(lhsOp backends.Op, lhsContractingAxes, lhsBatchAxes
 		return nil, err
 	}
 	lhs, rhs := inputs[0], inputs[1]
-	dtype := lhs.shape.DType
-	if dtype != rhs.shape.DType {
-		return nil, errors.Errorf("DotGeneral lhs (left-hand-side) and rhs operands don't match data types: %s and %s", dtype, rhs.shape.DType)
+	lhsDType := lhs.shape.DType
+	rhsDType := rhs.shape.DType
+
+	if lhsDType != rhsDType {
+		return nil, errors.Errorf("DotGeneral lhs (left-hand-side) and rhs operands don't match data types: %s and %s", lhsDType, rhsDType)
+	}
+
+	dtype := lhsDType
+
+	// Determine the final output dtype for the node.
+	// For int8/uint8: output is int32 (widened accumulation type that's also the final output).
+	// For bfloat16/float16: output remains the input dtype (internal accumulation uses float32 but converts back).
+	nodeOutputDType := dtype
```

Comment:
So this is dangerous. 

The DotGeneral API specificed in `backends.Backend` doesn't include any automatic promotion of the dtype (dtypes.Int8 to dtypes.In32), so this would break it -- let's say for folks who don't want the promotion.

I modeled the `backends.Backend` interface after the StableHLO's one (see [DotGeneral there](https://openxla.org/stablehlo/spec#dot_general)). But it doesn't need to be identical ... In StableHLO, except for quantized types, the output dtype must be equal to the input dtype (but there is the option to set the accumulating dtype).

Let's VC about what we want to do here ? I'll try to study more about quantized types in StableHLO/XLA.

Reply:
Good catch. Let's discuss this more tomorrow, but now, here's what I did: 
 * removed the automatic `int8 --> int32` promotion for the output, so the output `dtype` now matches the input `dtype`
 * added `int32` accumulation internally to prevent overflow, then saturate final results to int8 range ([-128, 127])
 * this is compatible with ONNX as onnx-gomlx handles the int32 promotion already https://github.com/gomlx/onnx-gomlx/blob/main/onnx/ops.go#L2016

Reply:
The `int32` also potentially has a cost (in the odd case for instance if the user is doing matmul of some sparse matrices, and they know it won't overflow). I wonder if we should add an accumulation dtype parameter to the generic DotGeneral (in `backends.Backend`) and let the user decide ?